nestjs/src/courses/team-distribution/team.controller.ts

```diff
@@ -0,0 +1,169 @@
+  @ApiOperation({ operationId: 'getTeamPassword' })
+  @RequiredRoles([CourseRole.Student, Role.Admin, CourseRole.Manager])
+  public async getTeamPassword(
+    @Req() req: CurrentRequest,
+    @Param('courseId', ParseIntPipe) courseId: number,
+    @Param('distributionId', ParseIntPipe) _: number,
+    @Param('id', ParseIntPipe) id: number,
+  ) {
+    const team = await this.teamService.findById(id);
+    const studentId = req.user.courses[courseId]?.studentId;
+    const isManager = req.user.isAdmin || req.user.courses[courseId]?.roles.includes(CourseRole.Manager);
+
+    if (studentId !== team.teamLeadId && !isManager) throw new BadRequestException();
+    return new TeamPasswordDto(team);
+  }
+
+  @Post('/:id/join')
+  @UseGuards(RoleGuard)
+  @ApiOkResponse({ type: TeamDto })
+  @ApiOperation({ operationId: 'joinTeam' })
+  @RequiredRoles([CourseRole.Student])
+  public async joinTeam(
+    @Req() req: CurrentRequest,
+    @Param('courseId', ParseIntPipe) courseId: number,
+    @Param('distributionId', ParseIntPipe) distributionId: number,
+    @Param('id', ParseIntPipe) id: number,
+    @Body() dto: JoinTeamDto,
+  ) {
+    const studentId = req.user.courses[courseId]?.studentId;
+    if (!studentId) throw new BadRequestException();
+    const team = await this.teamService.findById(id);
+    const student = await this.studentService.getStudentDetailed(studentId);
+    if (team.teamDistribution.strictStudentsCount && team.teamDistribution.studentsCount <= team.students.length + 1) {
+      throw new BadRequestException();
+    }
+    if (student.teams.find(t => t.teamDistributionId === team.teamDistributionId)) throw new BadRequestException();
+    if (!student.teamDistribution.find(td => td.id === distributionId)) throw new BadRequestException();
+    if (dto.password !== team.password) throw new BadRequestException('Invalid password');
+    await this.studentService.addStudentToTeam(studentId, team);
+    await this.studentService.deleteStudentFromTeamDistribution(studentId, team.teamDistribution);
+    return new TeamDto(team);
+  }
+
+  @Post('/:id/leave')
+  @UseGuards(RoleGuard)
+  @ApiOkResponse()
+  @ApiOperation({ operationId: 'leaveTeam' })
+  @RequiredRoles([CourseRole.Student])
+  public async leaveTeam(
+    @Req() req: CurrentRequest,
+    @Param('courseId', ParseIntPipe) courseId: number,
+    @Param('distributionId', ParseIntPipe) distributionId: number,
+    @Param('id', ParseIntPipe) id: number,
+  ) {
+    const studentId = req.user.courses[courseId]?.studentId;
+    if (!studentId) throw new BadRequestException();
+    await this.studentService.deleteStudentFromTeam(studentId, id);
+    const teamDistribution = await this.distributionService.getById(distributionId);
+    await this.studentService.addStudentToTeamDistribution(studentId, teamDistribution, false);
+    const { studentsCount } = await this.teamService.getStudentsCountInTeam(id);
+    if (studentsCount === '0') await this.teamService.remove(id);
```

Comment:
seams bit strange that service returns `string` count, wouldn't it be better to return `number`?

Reply:
fixed